src/coreclr/jit/lsrabuild.cpp

```diff
@@ -1456,6 +1456,34 @@ Interval* LinearScan::getUpperVectorInterval(unsigned varIndex)
     unreached();
 }
 
+//------------------------------------------------------------------------
+// shouldBuildUpperVectorSaveRestore - Check if vector save/restore is needed.
+//
+// Arguments:
+//    tree       - The current node being handled
+//
+// Notes: Vector save/restore is usually not needed before the call that never returns.
+//        It is also not needed if the block ends with a throw.
+bool LinearScan::shouldBuildUpperVectorSaveRestore(GenTree* tree)
+{
+    if ((tree != nullptr) && tree->IsCall())
+    {
+        if (tree->AsCall()->IsNoReturn())
+        {
+            // No point in having vector save/restore if the call will not return.
```

Comment:
It's not clear to me why this would be safe. Couldn't you have computation after a call that requires restoring the registers, and before a throw?

Reply:
Seems like it's only safe if there are not uses of the (restored) register between this point and the throw, in which case the restore is "dead code".

Reply:
Vector restore happens at two places: 
- At the end of every block, whenever any `largeVectorVars` was partially spilled
- At the locations where we see the use

While it should be safe to eliminate the restores (and hence the saves) that happen for #1 above if the block is a THROW block, but there can be a scenario (as you pointed out) where eliminating save/restore will be wrong:

```asm
save(Vx)  ; <- shouldn't be eliminated because of the use
CALL ...
Vx = restore() ;  <- shouldn't be eliminated because of the use
 = use(Vx)
...
THROW
```

Since UpperVectorSave refpositions are created when we build kills, it is hard to look ahead to see which ones of them will be used (and hence restored) and thus can be eliminated. For e.g. in following case, we can (should) eliminate the unnecessary save/restore.

```asm
save(Vx)  ; <- safe to eliminate
CALL ...
Vx = restore() ;  <- safe to eliminate
...
THROW
```

To summarize, I was trying to eliminate vector saves, but it might not be safe (as Bruce pointed), so I will just try to at least eliminate the restore that happens at the end of the THROW block.

Reply:
Hhm, looks like our model is not built such that it can withstand to see only saves and not (optimally removed) restores.

Reply:
Pushed an attempt to track for a block that throws, if a spill done before the call will ever be used. If not, skip save/restore for them. While I am at it, I will check what it takes to fix the problem @sandreenko pointed out in https://github.com/dotnet/runtime/pull/59315.